sql: stop creating indexes on the origin side of an FK

rohany opened the pull request:
Release note (sql change): Indexes are no longer required on the origin
side of a foreign key relationship, and are no longer automatically
created.

rohany committed:
sql: stop creating indexes on the origin side of an FK

Release note (sql change): Indexes are no longer required on the origin
side of a foreign key relationship, and are no longer automatically
created.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50771)
<!-- Reviewable:end -->

RaduBerinde commented:
Thanks for taking this on! We need to check how this affects TPCC. An easy way to do that would be to just drop the FK indexes and see how the performance changes. If it turns out to be better, we'll need to update the backups/fixtures we use to test (unfortunately, the new ones won't work on previous releases). If it turns out to be worse, we'll need to update the TPCC workload to add the indexes explicitly.

rohany commented:
Yeah, @awoods187 gave me a heads up to test that. Once this PR gets into a better state (and #50779 is resolved) I'll play around with TPCC.

rohany commented:
Friendly ping!

rytaft reviewed (commented):
> Reviewed 1 of 20 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50771)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang, @RaduBerinde, and @rohany)

___
*[pkg/sql/opt/xform/custom_funcs.go, line 1781 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50771#-MBff3Pb2ygfTYHRT9vy:-MBff3Pb2ygfTYHRT9vz:b-6p3nup) ([raw file](https://github.com/cockroachdb/cockroach/blob/640e601d7db31eb7e41e1049eccb1655409a6f10/pkg/sql/opt/xform/custom_funcs.go#L1781)):*
> ```Go
> 		}
> 
> 		if joinType == opt.SemiJoinOp || joinType == opt.AntiJoinOp {
> ```

This is a duplicate of the block above


<!-- Sent from Reviewable.io -->

rohany commented:
> This is a duplicate of the block above

Woops, dirt from a rebase.

rohany commented:
Pinging again!

rytaft commented:
cc @raduberinde @lucy-zhang 





<!-- Sent from Reviewable.io -->

thoszhang reviewed (approved):
Should we do any 20.1/20.2 mixed-version testing to ensure that the restrictions still exist? We could add a logic test configuration like we've done in the last few releases.

> Reviewed 18 of 20 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50771#-:-MCCxPn68-M64LAlfIYe:bo98d3x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rohany)

___
*[pkg/sql/create_table.go, line 639 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50771#-MCCwZHm5nK6afW6Hty8:-MCCwZHm5nK6afW6Hty9:b-7vaefa) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bf2f823190028b3f4ad9c28363e5863b0f0a330/pkg/sql/create_table.go#L639)):*
> ```Go
> 		}
> 		// Ensure that the origin columns don't have duplicates.
> 		if _, ok := originColMap[col.Name]; ok {
> ```

Why is the deduplication here by name instead of ID?


<!-- Sent from Reviewable.io -->

rohany commented:
> Should we do any 20.1/20.2 mixed-version testing to ensure that the restrictions still exist? We could add a logic test configuration like we've done in the last few releases.

Good idea, added.

> Why is the deduplication here by name instead of ID?

No good reason. Moved to ID.

rohany commented:
bors r=lucy-zhang

RaduBerinde commented:
Awesome!

ericharmeling commented:
So this change makes it possible for a table to have a foreign key constraint, but no index on the constrained columns, correct?

If so, it looks like we need to update some cockroach workload schemas to reflect this change. Specifically, `tpcc` and `movr`, two schemas that we use extensively in documentation examples on foreign keys, optimization, partitioning, indexing, and other important topics.

## `tpcc`

```
root@:26257/tpcc> select distinct on (table_name, index_name) table_name, index_name from [show indexes from database tpcc];
  table_name |    index_name
-------------+--------------------
  warehouse  | primary
  district   | primary
  customer   | primary
  customer   | customer_idx
  history    | primary
  order      | primary
  order      | order_idx
  new_order  | primary
  item       | primary
  stock      | primary
  stock      | stock_item_fk_idx
  order_line | primary
(12 rows)
```

```
root@:26257/tpcc> select table_name, referenced_table_name from information_schema.referential_constraints;
  table_name | referenced_table_name
-------------+------------------------
  district   | warehouse
  customer   | district
  history    | customer
  history    | district
  order      | customer
  new_order  | order
  stock      | warehouse
  stock      | item
  order_line | order
  order_line | stock
(10 rows)
```

```
root@:26257/tpcc> show create stock;
  table_name |                                                                                             create_statement
-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  stock      | CREATE TABLE public.stock (
             |     s_i_id INT8 NOT NULL,
             |     s_w_id INT8 NOT NULL,
             |     s_quantity INT8 NULL,
             |     s_dist_01 CHAR(24) NULL,
             |     s_dist_02 CHAR(24) NULL,
             |     s_dist_03 CHAR(24) NULL,
             |     s_dist_04 CHAR(24) NULL,
             |     s_dist_05 CHAR(24) NULL,
             |     s_dist_06 CHAR(24) NULL,
             |     s_dist_07 CHAR(24) NULL,
             |     s_dist_08 CHAR(24) NULL,
             |     s_dist_09 CHAR(24) NULL,
             |     s_dist_10 CHAR(24) NULL,
             |     s_ytd INT8 NULL,
             |     s_order_cnt INT8 NULL,
             |     s_remote_cnt INT8 NULL,
             |     s_data VARCHAR(50) NULL,
             |     CONSTRAINT "primary" PRIMARY KEY (s_w_id ASC, s_i_id ASC),
             |     CONSTRAINT fk_s_w_id_ref_warehouse FOREIGN KEY (s_w_id) REFERENCES public.warehouse(w_id),
             |     CONSTRAINT fk_s_i_id_ref_item FOREIGN KEY (s_i_id) REFERENCES public.item(i_id),
             |     INDEX stock_item_fk_idx (s_i_id ASC),
             |     FAMILY "primary" (s_i_id, s_w_id, s_quantity, s_dist_01, s_dist_02, s_dist_03, s_dist_04, s_dist_05, s_dist_06, s_dist_07, s_dist_08, s_dist_09, s_dist_10, s_ytd, s_order_cnt, s_remote_cnt, s_data)
             | )
(1 row)
```

Ostensibly, `stock_item_fk_idx` is the only index created expressly for the purpose of indexing fk-constrained columns. Do we need this index for anything else? If not, we should remove it, or, at the very least, we should change the name of `stock_item_fk_idx` to `stock_idx`, to be consistent with the other indexes. (Unless of course this is standard TPCC stuff that I don't know about)

If the fk index does, in fact, improve performance, then we should offer guidance to users on how to determine when to add an index on an fk-constrained column and when not to. Thoughts?

## `movr`

```
root@:26257/movr> select distinct on (table_name, index_name) table_name, index_name from [show indexes from database movr];
          table_name         |                  index_name
-----------------------------+------------------------------------------------
  users                      | primary
  vehicles                   | primary
  vehicles                   | vehicles_auto_index_fk_city_ref_users
  rides                      | primary
  rides                      | rides_auto_index_fk_city_ref_users
  rides                      | rides_auto_index_fk_vehicle_city_ref_vehicles
  vehicle_location_histories | primary
  promo_codes                | primary
  user_promo_codes           | primary
(9 rows)

root@:26257/movr> select table_name, referenced_table_name from information_schema.referential_constraints;
          table_name         | referenced_table_name
-----------------------------+------------------------
  vehicles                   | users
  rides                      | users
  rides                      | vehicles
  vehicle_location_histories | rides
  user_promo_codes           | users
(5 rows)

```

The auto-generated foreign key indexes are still on `rides` and `vehicles` (but for some reason are missing from `vehicle_location_histories` and `user_promo_codes`?). We should remove these indexes if they no longer serve a purpose. If we want to keep them, we should rename them.

rohany commented:
> So this change makes it possible for a table to have a foreign key constraint, but no index on the constrained columns, correct?

Yep

The remaining stock index on the TPCC schema was kept because it seemed like it was beneficial to have for performance.

To decide whether such an index is needed, we can direct users to looking at the plans generated for queries involving FK checks. If the joins or scans produced there would benefit from an index then the index could be used. Alternatively, if an index is already there for an FK, and an examination of the query plans show that the index either isn't used, or could be replaced with another existing index, then a user could drop that index.

I never got around to updating the movr schema's foreign keys.

RaduBerinde commented:
I think we should rename the TPCC stock index, the index is beneficial but not because of foreign keys.

> If the fk index does, in fact, improve performance, then we should offer guidance to users on how to determine when to add an index on an fk-constrained column and when not to. Thoughts?

Yes, we need to provide some guidance. If we are frequently deleting parent rows, or updating the values of referenced columns in the parent, the index is usually beneficial.

Postgres says this ([link](https://www.postgresql.org/docs/current/ddl-constraints.html)]:
> A foreign key must reference columns that either are a primary key or form a unique constraint. This means that the referenced columns always have an index (the one underlying the primary key or unique constraint); so checks on whether a referencing row has a match will be efficient. Since a DELETE of a row from the referenced table or an UPDATE of a referenced column will require a scan of the referencing table for rows matching the old value, it is often a good idea to index the referencing columns too. Because this is not always needed, and there are many choices available on how to index, declaration of a foreign key constraint does not automatically create an index on the referencing columns.

lucy-zhang commented:
> The auto-generated foreign key indexes are still on rides and vehicles (but for some reason are missing from `vehicle_location_histories` and `user_promo_codes`?).

I think the primary keys of `vehicle_location_histories` and `user_promo_codes` were taking care of that requirement.

lucy-zhang commented:
I just opened https://github.com/cockroachdb/cockroach/issues/54764 to track this for movr and put it on the Features board.

Bessonov commented:
This one is very sneaky. I had been spending many hours on optimization of our software, queries, creating indices etc... and discovered that the new version of CRDB doesn't creates indices for FKs anymore... :(

I think this part of [the documentation](https://www.cockroachlabs.com/docs/v21.1/foreign-key.html) was forgotten:
> If an index on the referenced column does not already exist, CockroachDB automatically creates one.

ericharmeling commented:
@Bessonov 
> If an index on the referenced column does not already exist, CockroachDB automatically creates one.

I will remove this. Good catch!

@knz 

> our FK <-> index mapping has changed in the latest version

The FK <-> index mapping has changed in 21.1? Or are you referring to the changes in this PR (i.e., in 20.2)?

knz commented:
> Or are you referring to the changes in this PR (i.e., in 20.2)?

I'm not sure which version things change. need to ask some schema

ericharmeling commented:
> I'm not sure in which version things have changed.

Pretty sure it's 20.2. (Here's my original docs PR: https://github.com/cockroachdb/docs/pull/8480).

Here's the fix: https://github.com/cockroachdb/docs/pull/10681
